core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java

```diff
@@ -214,59 +216,82 @@ PolicySelection decideLoadBalancerProvider(
         }
       }
 
-      if (haveBalancerAddress) {
-        // This is a special case where the existence of balancer address in the resolved address
-        // selects "grpclb" policy regardless of the service config.
-        LoadBalancerProvider grpclbProvider = registry.getProvider("grpclb");
-        if (grpclbProvider == null) {
-          if (backendAddrs.isEmpty()) {
-            throw new PolicyException(
-                "Received ONLY balancer addresses but grpclb runtime is missing");
-          }
-          if (!roundRobinDueToGrpclbDepMissing) {
-            roundRobinDueToGrpclbDepMissing = true;
-            String errorMsg = "Found balancer addresses but grpclb runtime is missing."
-                + " Will use round_robin. Please include grpc-grpclb in your runtime depedencies.";
-            helper.getChannelLogger().log(ChannelLogLevel.ERROR, errorMsg);
-            logger.warning(errorMsg);
-          }
-          return new PolicySelection(
-              getProviderOrThrow(
-                  "round_robin", "received balancer addresses but grpclb runtime is missing"),
-              backendAddrs, null);
-        } else {
-          return new PolicySelection(grpclbProvider, servers, null);
-        }
-      }
-      roundRobinDueToGrpclbDepMissing = false;
-
       List<LbConfig> lbConfigs = null;
       if (config != null) {
         List<Map<String, Object>> rawLbConfigs =
             ServiceConfigUtil.getLoadBalancingConfigsFromServiceConfig(config);
         lbConfigs = ServiceConfigUtil.unwrapLoadBalancingConfigList(rawLbConfigs);
       }
       if (lbConfigs != null && !lbConfigs.isEmpty()) {
+        // If there isn't a preselecbted, try to get the provider as configured
         LinkedHashSet<String> policiesTried = new LinkedHashSet<>();
         for (LbConfig lbConfig : lbConfigs) {
           String policy = lbConfig.getPolicyName();
-          LoadBalancerProvider provider = registry.getProvider(policy);
-          if (provider != null) {
-            if (!policiesTried.isEmpty()) {
-              helper.getChannelLogger().log(
-                  ChannelLogLevel.DEBUG,
-                  "{0} specified by Service Config are not available", policiesTried);
+          if (policy.equals(GRPCLB_POLICY_NAME)) {
```

Comment:
Get rid of this check here and instead do a check after `registry.getProvider(policy)`. The check will just swap out and use `servers` instead of `backendAddrs`.